lib/core/src/display/navigation/camera_manager.rs

```diff
@@ -0,0 +1,197 @@
+use crate::prelude::*;
+
+use super::mouse_manager::State;
+use super::mouse_manager::MouseManager;
+use super::mouse_manager::MouseClickEvent;
+use super::mouse_manager::MouseWheelEvent;
+use super::mouse_manager::MousePositionEvent;
+use super::mouse_manager::MouseButton;
+use crate::system::web::Result;
+use crate::display::rendering::DOMContainer;
+
+use nalgebra::Vector2;
+use std::rc::Rc;
+use std::cell::RefCell;
+use nalgebra::zero;
+
+// ====================
+// === ZoomingEvent ===
+// ====================
+
+/// A struct holding zooming event information.
+pub struct ZoomingEvent {
+    pub focus  : Vector2<f32>,
+    pub amount : f32
+}
+
+impl ZoomingEvent {
+    fn from_mouse_wheel(event:MouseWheelEvent, data:&Rc<RefCell<CameraManagerData>>) -> Self {
+        let amount = event.movement_y;
+        let focus  = data.borrow().mouse_position;
+        Self { amount, focus }
+    }
+
+    fn from_mouse_move(event:MousePositionEvent, focus:Vector2<f32>) -> Self {
+        let amount = event.position.y - event.previous_position.y;
+        Self { focus, amount }
+    }
+}
+
+// ====================
+// === PanningEvent ===
+// ====================
+
+/// A struct holding panning event information.
+pub struct PanningEvent {
+    pub movement : Vector2<f32>
+}
+
+impl PanningEvent {
+    fn from(event:MousePositionEvent) -> Self {
+        let mut movement   = event.position - event.previous_position;
+                movement.x = -movement.x;
+        Self { movement }
+    }
+}
+
+// =============
+// === Event ===
+// =============
+
+/// An enumeration representing navigation events.
+pub enum Event {
+    Zooming(ZoomingEvent),
+    Panning(PanningEvent)
+}
+
+// ====================
+// === MovementType ===
+// ====================
+
+#[derive(PartialEq)]
+enum MovementType {
+    Panning,
+    Zooming { focus : Vector2<f32> }
+}
+
+// ========================
+// === EventHandlerData ===
+// ========================
+
+struct CameraManagerData {
+    event_queue       : Vec<Event>,
+    movement_type     : Option<MovementType>,
+    mouse_position    : Vector2<f32>
+}
+
+impl Default for CameraManagerData {
+    fn default() -> Self {
+        let event_queue       = default();
+        let mouse_position    = zero();
+        let movement_type     = None;
+        Self { event_queue, mouse_position, movement_type }
+    }
+}
+
+// =====================
+// === CameraManager ===
+// =====================
+
+/// Struct used to handle mouse events such as MouseDown, MouseMove, MouseUp,
+/// MouseLeave and MouseWheel.
+pub struct CameraManager {
+    mouse_manager : MouseManager,
+    data          : Rc<RefCell<CameraManagerData>>
+}
```

Comment:
I would like not to use this pattern. Keeping one field behind shared reference and other not is error prone because it forces all functions which operate on `CameraManager` to handle the borrowing manually. Let's talk live for more details if needed.

Reply:
Beyond MouseManager, I am also keeping track of the listeners. They don't need to be shared, so that's why I am keeping them outside CameraManagerData